demos/instr.cc

```diff
@@ -114,3 +114,49 @@ uint64_t ReadLatency(const void *memory) {
   LFence();  // Necessary for x86 MSVC.
   return result;
 }
+
+#ifdef __GNUC__
+__attribute__((noinline))
+void UnwindStackAndSlowlyReturnTo(const void *address) {
+  asm volatile(
+#if defined(__x86_64__) || defined(_M_X64)
+      "addq $8, %%rsp\n"
+      "popstack:\n"
+      "addq $8, %%rsp\n"
+      "cmpq %0, (%%rsp)\n"
+      "jnz popstack\n"
+      "clflush (%%rsp)\n"
+      "mfence\n"
+      "lfence\n"
+#elif defined(__i386__) || defined(_M_IX86)
+      "addl $4, %%esp\n"
+      "popstack:\n"
+      "addl $4, %%esp\n"
+      "cmpl %0, (%%esp)\n"
+      "jnz popstack\n"
+      "clflush (%%esp)\n"
+      "mfence\n"
+      "lfence\n"
+#elif defined(__aarch64__)
+      // Unwind until the magic value and pop the magic value.
+      "movz x9, 0x4567\n"
+      "movk x9, 0x0123, lsl 16\n"
+      "movk x9, 0xba98, lsl 32\n"
+      "movk x9, 0xfedc, lsl 48\n"
+      "popstack:\n"
+      "ldr x10, [sp], #16\n"
+      "cmp x9, x10\n"
+      "bne popstack\n"
+      // Push the return address on the stack.
+      "str %0, [sp, #-16]!\n"
+      // Pop the return address slowly from the stack and return.
+      "mov x11, sp\n"
+      "dc civac, x11\n"
+      "dsb sy\n"
+      "ldr x30, [sp], #16\n"
+#else
+#  error Unsupported CPU.
+#endif
+      "ret\n"::"r"(address));
```

Comment:
It seems a bit too clever to factor out `ret` just because it happens to be an instruction in each architecture. I would actually prefer the `asm volatile` block be replicated within each branch of the `#ifdef`.

Reply:
Done.